rescript/ncbi.py

```diff
@@ -0,0 +1,141 @@
+        else:
+            seqs, taxids = get_nuc_for_accs(accs)
+
+    taxa = get_taxonomies(taxids)
+
+    seqs = DNAIterator(DNA(v, metadata={'id': k}) for k, v in seqs.items())
+    taxa = DataFrame(taxa, index=['Taxon']).T
+    taxa.index.name = 'Feature ID'
+
+    return seqs, taxa
+
+
+def get_from_post(ids, params):
+    assert len(ids) >= 1, "need at least one id"
+    epost = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi'
+    efetch = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi'
+    data = {'db': params['db'], 'id': ','.join(ids)}
+    r = requests.post(epost, data=data)
+    r.raise_for_status()
+    webenv = parse(r.content)['ePostResult']
+    params['WebEnv'] = webenv['WebEnv']
+    params['query_key'] = webenv['QueryKey']
+    # Need to better handle the case of invalid ids
+    # Need to better handle intermittent fault where we
+    #  don't get the whole data set back
+    data = []
+    while len(data) < len(ids):
+        params['retstart'] = len(data)
+        r = requests.get(efetch, params=params)
+        r.raise_for_status()
+        chunk = parse(r.content)
+        chunk = list(chunk.values()).pop()
+        chunk = list(chunk.values()).pop()
+        if len(ids) == 1:
+            return [chunk]
+        data.extend(chunk)
+    return data
+
+
+def get_nuc_for_accs(accs):
+    params = dict(
+        db='nuccore', rettype='fasta', retmode='xml'
+    )
+    records = get_from_post(accs, params)
+    seqs = {}
+    taxids = {}
+    for rec in records:
+        seqs[rec['TSeq_accver']] = rec['TSeq_sequence']
+        taxids[rec['TSeq_accver']] = rec['TSeq_taxid']
+    return seqs, taxids
+
+
+def get_nuc_for_query(query):
+    esearch = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi'
+    efetch = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi'
+    params = dict(
+        db='nuccore', term=query, usehistory='y', retmax=0
+    )
+    r = requests.get(esearch, params=params)
+    r.raise_for_status()
+    webenv = parse(r.content)['eSearchResult']
+    params = dict(
+        db='nuccore', rettype='fasta', retmode='xml',
+        WebEnv=webenv['WebEnv'], query_key=webenv['QueryKey']
+    )
+    seqs = {}
+    taxids = {}
+    while len(seqs) < int(webenv['Count']):
+        params['retstart'] = len(seqs)
+        r = requests.get(efetch, params=params)
+        r.raise_for_status()
+        data = parse((r.content))
+        data = list(data.values()).pop()
+        data = list(data.values()).pop()
+        for rec in data:
+            seqs[rec['TSeq_accver']] = rec['TSeq_sequence']
+            taxids[rec['TSeq_accver']] = rec['TSeq_taxid']
+    return seqs, taxids
+
+
+def get_taxonomies(taxids):
+    params = dict(db='taxonomy')
+    ids = list(map(str, taxids.values()))
+    records = get_from_post(ids, params)
+    taxa = {}
+    levels = ['kingdom', 'phylum', 'class', 'order', 'family', 'genus',
+              'species']
```

Comment:
We should likely grab all taxonomies and then subset / propagate them afterwards. We also applied this approach to the [SILVA parsing](https://github.com/bokulich-lab/RESCRIPt/blob/772294cd0a9fed069f2b2dfcd0f1d6144722eff3/rescript/parse_silva_taxonomy.py#L23).  

I suggest this, as many eukaryote data in GenBank do not have , e.g. an "Order" or "Family" but have "Sub-Order" or "Super-Family"... or some other rank. This was always an issue for me when building databases for other marker genes, like 12S, COI, trnL. Especially when plants have "Divisions", "Tribes", and "Varieties". 

Anyway... this is all assuming we want to open this up beyond microbial taxonomy. I figured, it should be on our radar, even if only for a later release.  :-)

Reply:
I agree that this needs to be configurable in some way. I am also happy to leave it to later.

As a first step, I have made "levels" an input parameter.

We could also add an "allowed-levels" an input parameter so that we could provide your silva level extrapolation algorithm.

To help me understand your algorithm, do you build a taxonomic tree based on "allowed-levels", of which the "levels" are a subset, then propagate labels down branches until all levels in "levels" are filled?

Reply:
Hi Ben,

The literal taxonomy tree was provided by SILVA. I just traversed the tree and pulled the taxids from all the nodes. Then used them to lookup the taxonomy and rank information from the other files they provided. However, this is not unlike what I've done with the old pycogent NCBI taxonomy parser code. I just traversed the hierarchy and either pulled everything.. or only the ranks I wanted and continued on. I can e-mail you some *rather crappy* bits of code (python 2.7) that I wrote to fetch genbank taxonomy from local blast output, based on 'alllowed ranks'.

Reply:
> you build a taxonomic tree based on "allowed-levels", of which the "levels" are a subset, then propagate labels down branches until all levels in "levels" are filled

Sorry, I forgot to address that. Essentially yes. Though in this case I was copying and propagating within a dataframe. But your cae case. I'd recommend propagation BEFORE extracting the levels you want. This way we would not have to worry about empty 'Family' or other empty ranks that might turn up. That is, the taxonomy would be propagated before you'd subset that information. This is what we are doing with SILVA via the dataframe.

Reply:
Thanks Mike for all your feedback (and chats offline). I took this algorithm and implemented in my taxonomy parsing code. I hope I got it right!